### PR TITLE
addcol/delcol add to an on disk ctable

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

**Comment by esc** [comment]
Thanks! I'll need to try it out but adding to the 0.8.1 release for the time being

**Comment by esc** [comment]
Thanks for the rebase! Much appreciated! I've added this to the 0.9.0 milestone as indicated on the mailinglist.

File: bcolz/ctable.py [hunk 1]
```diff
@@ -516,7 +516,13 @@ def delcol(self, name=None, pos=None):
             name = self.names[pos]
 
         # Remove the column
-        self.cols.pop(name)
+        col = self.cols.pop(name)
+
+        # remove the data if we have a rootdir
```
Comment (esc) [inline bcolz/ctable.py]: I am thinking the carray needs to know how to delete itself. Seperation of concerns.

**Comment by esc** [comment]
Regarding the deletion itself, it is possible to create a column that lives in-memory:

```
In [5]: t
Out[5]: 
ctable((10,), [('f0', '<i4'), ('f1', '<f8'), ('f2', 'S2')])
  nbytes: 140; cbytes: 48.00 KB; ratio: 0.00
  cparams := cparams(clevel=5, shuffle=True, cname='blosclz')
  rootdir := 'A'
[(0, 0.0, 's0') (1, 2.0, 's1') (2, 4.0, 's2') (3, 6.0, 's3') (4, 8.0, 's4')
 (5, 10.0, 's5') (6, 12.0, 's6') (7, 14.0, 's7') (8, 16.0, 's8')
 (9, 18.0, 's9')]

In [6]: t.delcol('f2')

In [7]: t.addcol(c, 'f2', rootdir=None)

In [8]: t
Out[8]: 
ctable((10,), [('f0', '<i4'), ('f1', '<f8'), ('f2', 'S2')])
  nbytes: 140; cbytes: 48.00 KB; ratio: 0.00
  cparams := cparams(clevel=5, shuffle=True, cname='blosclz')
  rootdir := 'A'
[(0, 0.0, 's0') (1, 2.0, 's1') (2, 4.0, 's2') (3, 6.0, 's3') (4, 8.0, 's4')
 (5, 10.0, 's5') (6, 12.0, 's6') (7, 14.0, 's7') (8, 16.0, 's8')
 (9, 18.0, 's9')]

In [9]: ls -lah A
total 24K
drwx------  4 esc esc 4.0K Jan 21 23:34 ./
drwx------ 12 esc esc 4.0K Jan 21 23:34 ../
-rw-------  1 esc esc    3 Jan 21 23:34 __attrs__
drwx------  4 esc esc 4.0K Jan 21 23:34 f0/
drwx------  4 esc esc 4.0K Jan 21 23:34 f1/
-rw-------  1 esc esc   30 Jan 21 23:35 __rootdirs__

In [10]: cat A/__rootdirs__ 
{"names": ["f0", "f1", "f2"]}
```

This raises the question if this should even be allowed or useful.

File: bcolz/ctable.py [hunk 2]
```diff
@@ -456,13 +455,14 @@ def addcol(self, newcol, name=None, pos=None, **kwargs):
         if len(newcol) != self.len:
             raise ValueError("`newcol` must have the same length than ctable")
 
-        if isinstance(newcol, np.ndarray):
-            if 'cparams' not in kwargs:
-                kwargs['cparams'] = self.cparams
+        if self.rootdir is not None:
```
Comment (esc) [inline bcolz/ctable.py]: I agree, adding columns to an existing on-disk table should make the carray also on-disk unless otherwise specified.

**Comment by esc** [comment]
Another thought crossed my mind, should delcol have the option of keeping the data on disk? If so, what should the API look like? I could imagine a keyword argument:

```
t.delcol('f2', keep=True)
```

But not sure what the correct default should be.

Alternatively we could have different methods:

```
t.purge('f2') # delete data also
t.purgecol('f2')
t.close('f2') # just flush and release the IOBuffer
t.closecol('f2')
```

Of course that would also mean, we need to be able to check for carrays that are within the rootdir, but not currently open and alsmo more methods, for example to close/open multiple columns etc.. But I think that decision can be deferred until later.

**Comment by CarstVaartjes** [comment]
Hi, i think the default should be deleting it (it's at least what I would expect as behaviour; you would expect the space to be freed up, when removing it, normally at least)

**Comment by esc** [comment]
I implemented the ``carray.purge`` in #130

**Comment by esc** [comment]
So changes would be, adding a ``keep`` keyword arg and handing off deletion responsibility to ``carray`` via ``purge``.

**Comment by esc** [comment]
I'll take care of it sometime this week (unless someone beats me to it!)

**Comment by esc** [comment]
Rebased and upgraded at #151

**Comment by esc** [comment]
Closing as #151 has been merged.

**Pull request closed without merging**